src/layout/umap.c

```diff
@@ -0,0 +1,1038 @@
+    }
+
+    phi = 1. / (1. + a * pow(dsq, b));
+    force = - mu * (2 * a * b * pow(dsq, b - 1)) * phi;
+    for (int d = 0; d != ndim; d++) {
+        VECTOR(*forces)[d] = force * VECTOR(*delta)[d];
+        /* clip force to avoid too rapid change */
+        igraph_i_umap_clip_force(&(VECTOR(*forces)[d]), 3);
+
+#ifdef UMAP_DEBUG
+        printf("force attractive: delta[%d] = %g, forces[%d] = %g\n", d, VECTOR(*delta)[d], d, VECTOR(*forces)[d]);
+#endif
+    }
+
+    return IGRAPH_SUCCESS;
+}
+
+/*xd is difference in x direction, mu is a weight */
+static igraph_error_t igraph_i_umap_repel(igraph_vector_t *delta, igraph_real_t mu,
+       igraph_real_t a, igraph_real_t b, igraph_vector_t *forces)
+{
+    igraph_real_t dsq, force;
+    igraph_real_t min_dist = 0.01;
+    igraph_integer_t ndim = igraph_vector_size(delta);
+
+    dsq = 0;
+    for (int d = 0; d != ndim; d++) {
+        dsq += VECTOR(*delta)[d] * VECTOR(*delta)[d];
+    }
+    dsq = fmax(min_dist * min_dist, dsq);
+
+    /* NOTE: in practice, in negative sampling mu is always zero because we
+     * *assume* the sample to be negative i.e. never a true edge */
+    force = (1 - mu) * (2 * b) / dsq / (1 + a * pow(dsq, b));
+    for (int d = 0; d != ndim; d++) {
+        VECTOR(*forces)[d] = force * VECTOR(*delta)[d];
+
+        /* clip force to avoid too rapid change */
+        igraph_i_umap_clip_force(&(VECTOR(*forces)[d]), 3);
+
+#ifdef UMAP_DEBUG
+        printf("force repulsive: delta[%d] = %g, forces[%d] = %g\n", d, VECTOR(*delta)[d], d, VECTOR(*forces)[d]);
+#endif
+    }
+
+    return IGRAPH_SUCCESS;
+}
+
+static igraph_error_t igraph_i_umap_apply_forces(const igraph_t *graph,  const igraph_vector_t *umap_weights,
+       igraph_matrix_t *layout, igraph_real_t a, igraph_real_t b, igraph_real_t prob,
+       igraph_real_t learning_rate, igraph_bool_t avoid_neighbor_repulsion)
+{
+    igraph_integer_t no_of_nodes = igraph_matrix_nrow(layout);
+    igraph_integer_t ndim = igraph_matrix_ncol(layout);
+    igraph_integer_t no_of_edges = igraph_ecount(graph);
+    igraph_integer_t from, to;
+    igraph_vector_t from_emb, to_emb, delta, forces;
+    /* The following is only used for small graphs, to avoid repelling your neighbors
+     * For large sparse graphs, it's not necessary. For large dense graphs, you should
+     * not be doing UMAP.
+     * */
+    igraph_vector_int_t neis, negative_vertices;
+
+    /* Initialize vectors */
+    IGRAPH_VECTOR_INIT_FINALLY(&from_emb, ndim);
+    IGRAPH_VECTOR_INIT_FINALLY(&to_emb, ndim);
+    IGRAPH_VECTOR_INIT_FINALLY(&delta, ndim);
+    IGRAPH_VECTOR_INIT_FINALLY(&forces, ndim);
+
+    if (avoid_neighbor_repulsion) {
+        IGRAPH_VECTOR_INT_INIT_FINALLY(&neis, 0);
+    }
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&negative_vertices, 0);
+
+    igraph_integer_t n_random_vertices = sqrt(no_of_nodes);
+
+    /* iterate over a random subsample of edges */
+    for (igraph_integer_t eid = 0; eid < no_of_edges; eid++) {
```

Comment:
We have a function to select a random subset of integers (or edge IDs), see `igraph_random_sample()`. Can you use it here? That should be more efficient for small `prob`. What are typical values of `prob`?

Reply:
that function allocates a vector of integers, which is not needed in this version. Since prob can be high if the user wants, I do not want to allocate a huge vector for no reason, hence the current version

Reply:
Again, it's best to add a comment about this.

Reply:
I have added a comment with Fabio's explanation. I've also added a TODO note mentioning that we should at least investigate using a geometric distribution to generate the next edge ID from the current if `prob` is small, similarly to how we do it in the G(n, p) Erdős-Rényi random graph model.